src/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -210,269 +211,245 @@ public sealed partial class TokenCache : ITokenCacheInternal
+                    requestParams.RequestContext.TelemetryCorrelationId,
+                    CacheEvent.TokenTypes.APPMETADATA,
+                    requestParams.Account,
+                    () => allAppMetadata = _accessor.GetAllAppMetadata())
+                .ConfigureAwait(false);
 
-            //TODO: bogavril - is the env ok here? Can I cache it or pass it in?
-            MsalAppMetadataCacheItem appMetadata;
-            await _semaphoreSlim.WaitAsync().ConfigureAwait(false);
-            try
-            {
-                await OnBeforeAccessAsync(args).ConfigureAwait(false);
+            var instanceMetadata = await ServiceBundle.InstanceDiscoveryManager.GetMetadataEntryTryAvoidNetworkAsync(
+                    requestParams.AuthorityInfo.CanonicalAuthority,
+                    allAppMetadata.Select(m => m.Environment),
+                    requestParams.RequestContext)
+                .ConfigureAwait(false);
 
-                appMetadata =
-                    instanceDiscoveryMetadata.Aliases
-                    .Select(env => _accessor.GetAppMetadata(new MsalAppMetadataCacheKey(ClientId, env)))
-                    .FirstOrDefault(item => item != null);
-
-                await OnAfterAccessAsync(args).ConfigureAwait(false);
-            }
-            finally
-            {
-                _semaphoreSlim.Release();
-            }
+            var appMetadata =
+                instanceMetadata.Aliases
+                .Select(env => _accessor.GetAppMetadata(new MsalAppMetadataCacheKey(ClientId, env)))
+                .FirstOrDefault(item => item != null);
 
             if (appMetadata == null)
             {
-                logger.Warning("No app metadata found. Returning unkown");
+                logger.Warning("No app metadata found. Returning unknown");
                 return null;
             }
 
             return appMetadata.FamilyId == familyId;
         }
 
+        private async Task LoadFromCacheAsync(
+            string telemetryId, CacheEvent.TokenTypes type, IAccount account, Action loadingAction)
+        {
+            var cacheEvent = new CacheEvent(
+                CacheEvent.TokenCacheLookup,
+                telemetryId)
+            {
+                TokenType = type
+            };
+
+            using (ServiceBundle.TelemetryManager.CreateTelemetryHelper(cacheEvent))
+            {
+                TokenCacheNotificationArgs args = new TokenCacheNotificationArgs(this, ClientId, account, false);
+                await _semaphoreSlim.WaitAsync().ConfigureAwait(false);
+                try
+                {
+                    await OnBeforeAccessAsync(args).ConfigureAwait(false);
+
+                    loadingAction();
+
+                    await OnAfterAccessAsync(args).ConfigureAwait(false);
+                }
+                finally
+                {
+                    _semaphoreSlim.Release();
+                }
+            }
+        }
+
+        // TODO: no telemetry 
```

Comment:
> TODO: no telemetry [](start = 10, length = 19)

i assume you're tracking these todo's somewhere? :)

Reply:
assume away :) 

I'll have to create work items.

Reply:
hahaha...if you have an overarching work item for this feature, then just add it there? one of us can also pick it up too. :)

Reply:
I will review this as part of the work item to reduce the number of cache hits.